### PR TITLE
Fix: DOS-913 Websocket Reconnection Refactor

### PR DESCRIPTION
**Pull request by sam-causalens**
## Motivation and Context
* The existing logic for retrying when the websocket connection fails lead to it looping forever when trying to reconnect due to faulty logic that lead to multiple sockets being created and the number of attempts always being 0 so never hitting the max count.
* My previous changes had also not gone far enough in terms of updating the global socket variable that left this code in an odd state after a reconnection. For example, new connections did not get a proper heartbeat setup as this was done on the global level.

## Implementation Description
* To address this, I've moved all the socket connection logic to the initialize function, this is now where the socket is created and has all it's listeners attached. Calling this function will create a brand new socket everytime with all the correct handlers rather than the subset that was left after my previous changes.
* The logic around the onClose event of the websocket has been moved into the class itself and is no longer recursive. The reconnectCount is tracked on the class itself which makes it consistent across retries so the attempts correctly stop after 10seconds.
* The onClose logic now uses the revamped initialize function to ensure that the socket is setup in the exact same way on reconnect as it was before.
* The timeout is now properly respected between attempts and does not lead to an effective infinite loop of retry requests

## Any new dependencies Introduced
jest-websocket-mock

## How Has This Been Tested?
* A suite of unit tests around the client have been added using `jest-websocket-mock` that test all the reconnection and initialization logic work correctly.

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [x] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [ ] I have added/updated Documentation.
- [ ] I have updated the appropriate changelog with a line for my changes.

### CODE REVIEW

File: packages/dara-core/js/api/websocket.tsx [hunk 1]
```diff
@@ -258,48 +182,106 @@ export class WebSocketClient implements WebSocketClientInterface {
 
     closeHandler: () => void;
 
-    constructor(_socket: WebSocket, _token: string, _liveReload = false) {
-        this.socket = _socket;
+    _maxAttempts: number;
```
Comment (krzysztof-causalens) [inline packages/dara-core/js/api/websocket.tsx:185]: For private variables in JS you can use the [# prefix](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Private_properties)

File: packages/dara-core/js/api/websocket.tsx [hunk 2]
```diff
@@ -258,48 +182,106 @@ export class WebSocketClient implements WebSocketClientInterface {
...
+        this._reconnectCount = 0;
+        this._pingInterval = null;
+        this._maxAttempts = maxAttempts;
+
+        // Lastly call initialize to setup the socket properly
+        this.socket = this.initialize();
     }
 
-    initialize(): void {
+    initialize(): WebSocket {
```
Comment (krzysztof-causalens) [inline packages/dara-core/js/api/websocket.tsx:207]: I might be missing something but I think this implementation doesn't implement liveReload at all?
Could we add a test case to ensure `location.reload` is invoked on reconnect if `liveReload` is true?

**Review by krzysztof-causalens — CHANGES_REQUESTED** [review]
Looks good, nice find. My only issue is that the new implementation is missing the live reload feature as far as I can tell

File: packages/dara-core/js/api/websocket.tsx (hunk 2 above)
Comment (sam-causalens) [reply packages/dara-core/js/api/websocket.tsx:207]: Ah, good catch. Sorry I took that out for simplicity when testing everything but forgot to put it back. I'll readd and add a test.

Comment (sam-causalens) [reply packages/dara-core/js/api/websocket.tsx]: Should be sorted now, let me know if you can think of any issues.

**Review by krzysztof-causalens — APPROVED** [review]

**Review by patricia-causalens — APPROVED** [review]
LGTM

**Pull request merged**